design/open/branch_checks.md

```diff
@@ -0,0 +1,171 @@
+# Branch Checks
+
+## User Story
+
+In order to achieve CI/CD for data, data engineering teams would like to validate the data itself before allowing to merge into a production branch.
+These validations might take longer than a typical HTTP request.
+
+## Goals
+
+Currently, lakeFS hooks allow running `pre-*` type hooks as blocking operations: They execute in a manner that is tied to an HTTP request.
+This puts a pretty short time window for hooks to complete in.
+
+Distributed processing frameworks such as Apache Spark typically have a long startup time and might take minutes (or even hours) to complete.
+
+The main goal of this proposal is to allow such validations to occur with a very long timeout (configurable, but could be days).
+
+## Non-Goals
+
+Out of scope for this proposal is any collaboration features that are typically associated with PRs (that provide similar functionality to the one being proposed).
+
+## High Level design
+
+This proposal adds a `Check` primitive to lakeFS branches. a Check is similar to a hook: it calls out to some external component with information about a modification or event that occurred in the system.
+
+The user flow for checks is the following:
+
+1. configure checks
+1. ensure all *mandatory* checks have passed for the given branch's HEAD
```

Comment:
You are mentioning "mandatory" checks - are there also checks which are not mandatory? If so - where do we configure a check to be mandatory or optional?

Reply:
Mandatory checks are enforced by the destination branch.

I would suggest - and perhaps its a good idea to cover it in the proposal - that this would happen using a pre-merge hook.